{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] DataManager: Dataset already exists, skipping fetch\n"
     ]
    }
   ],
   "source": [
    "%run dataset.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = os.path.join(\"..\", \"res\", \"datasets\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dman = DataManager(\"mymail@example.com\", data_folder)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.DataManager at 0x18b517386a0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dataset_name = \"RNA Dataset\"\n",
    "queries = [\"RNA\", \"mRNA\", \"tRNA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO] DataManager: Dataset already exists, skipping fetch\n"
     ]
    }
   ],
   "source": [
    "dman.create_dataset(queries, dataset_name, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "abstracts = dman.load_full_dataset(dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[DEBUG] tensorflow: Falling back to TensorFlow client; we recommended you install the Cloud TPU client directly with pip install cloud-tpu-client.\n"
     ]
    }
   ],
   "source": [
    "from tensorflow.keras.preprocessing.text import Tokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer = Tokenizer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.utils import to_categorical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.models import Sequential"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.layers import Dense, LSTM, Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.preprocessing.sequence import pad_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Long noncoding RNA nuclear paraspeckle assembly transcript 1 (lncRNA NEAT1) is abnormally expressed in numerous tumors and functions as an oncogene, but the role of NEAT1 in laryngocarcinoma is largely unknown. Our study validated that NEAT1 expression was markedly upregulated in laryngocarcinoma tissues and cells. Downregulation of NEAT1 dramatically suppressed cell proliferation and invasion through inhibiting miR-524-5p expression. Additionally, NEAT1 overexpression promoted cell growth and metastasis, while overexpression of miR-524-5p could reverse the effect. NEAT1 increased the expression of histone deacetylase 1 gene (HDAC1) via sponging miR-524-5p. Mechanistically, overexpression of HDAC1 recovered the cancer-inhibiting effects of miR-524-5p mimic or NEAT1 silence by deacetylation of tensin homolog deleted on chromosome ten (PTEN) and inhibiting AKT signal pathway. Moreover, in vivo experiments indicated that silence of NEAT1 signally suppressed tumor growth. Taken together, knockdown of NEAT1 suppressed laryngocarcinoma cell growth and metastasis by miR-524-5p/HDAC1/PTEN/AKT signal pathway, which provided a potential therapeutic target for laryngocarcinoma.'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abstracts[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "str"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(abstracts[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_text(doc):\n",
    "    tokens = doc.split()\n",
    "    table =str.maketrans('','', string.punctuation)\n",
    "    tokens = [w.translate(table) for w in tokens]\n",
    "    tokens = [word for word in tokens if word.isalpha()]\n",
    "    tokens = [word.lower() for word in tokens]\n",
    "    return tokens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokens = clean_text(abstracts[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "134"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tokens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "94"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(tokens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "length = 30+1\n",
    "lines = []\n",
    "for i in range(length, len(tokens)):\n",
    "    seq = tokens[i-length:i]\n",
    "    line = ' '.join(seq)\n",
    "    lines.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "lines = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "length = 50+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "for abstract in abstracts:\n",
    "    tokens = clean_text(abstract)\n",
    "    for i in range(length, len(tokens)):\n",
    "        seq = tokens[i-length:i]\n",
    "        line = ' '.join(seq)\n",
    "        lines.append(line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2262"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Build LSTM Model and Prepare x and y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "tokenizer = Tokenizer()\n",
    "tokenizer.fit_on_texts(lines)\n",
    "sequences = tokenizer.texts_to_sequences(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences = np.array(sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "X, y = sequences[:, :-1], sequences[:, -1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1039,  289,   30,  746, 1018, 1009,  998,  988,   17,  969,  350,\n",
       "          4,  949,  850,    2,  919,   19,   34,  893,  186,    1,  116,\n",
       "          3,    4,  314,   17,  341,  336,   64,   23,  199,   15,    7,\n",
       "          6,  797,  197,    4,  314,   74,    2,   14,  747,    3,  733,\n",
       "        174,   25,   39,    2,   89,   53])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "204"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'the': 1,\n",
       " 'and': 2,\n",
       " 'of': 3,\n",
       " 'in': 4,\n",
       " 'to': 5,\n",
       " 'was': 6,\n",
       " 'expression': 7,\n",
       " 'were': 8,\n",
       " 'with': 9,\n",
       " 'a': 10,\n",
       " 'by': 11,\n",
       " 'for': 12,\n",
       " 'protein': 13,\n",
       " 'cells': 14,\n",
       " 'that': 15,\n",
       " 'this': 16,\n",
       " 'is': 17,\n",
       " 'p': 18,\n",
       " 'as': 19,\n",
       " 'we': 20,\n",
       " 'from': 21,\n",
       " 'levels': 22,\n",
       " 'study': 23,\n",
       " 'results': 24,\n",
       " 'cell': 25,\n",
       " 'nscs': 26,\n",
       " 'using': 27,\n",
       " 'showed': 28,\n",
       " 'gene': 29,\n",
       " 'rna': 30,\n",
       " 'increased': 31,\n",
       " 'genes': 32,\n",
       " 'analysis': 33,\n",
       " 'an': 34,\n",
       " 'transcription': 35,\n",
       " 'significantly': 36,\n",
       " 'on': 37,\n",
       " 'different': 38,\n",
       " 'proliferation': 39,\n",
       " 'or': 40,\n",
       " 'most': 41,\n",
       " 'which': 42,\n",
       " 'sourdough': 43,\n",
       " 'viability': 44,\n",
       " 'b': 45,\n",
       " 'identified': 46,\n",
       " 'mrna': 47,\n",
       " 'after': 48,\n",
       " 'potential': 49,\n",
       " 'compared': 50,\n",
       " 'decidualization': 51,\n",
       " 'dlnbp': 52,\n",
       " 'through': 53,\n",
       " 'related': 54,\n",
       " 'evaluated': 55,\n",
       " 'stable': 56,\n",
       " 'ribosomal': 57,\n",
       " 'treated': 58,\n",
       " 'trfs': 59,\n",
       " 'signal': 60,\n",
       " 'crayfish': 61,\n",
       " 'decreased': 62,\n",
       " 'viruses': 63,\n",
       " 'our': 64,\n",
       " 'function': 65,\n",
       " 'can': 66,\n",
       " 'their': 67,\n",
       " 'hypoxic': 68,\n",
       " 'reverse': 69,\n",
       " 'mm': 70,\n",
       " 'these': 71,\n",
       " 'methods': 72,\n",
       " 'also': 73,\n",
       " 'tissues': 74,\n",
       " 'brain': 75,\n",
       " 'clpfesa': 76,\n",
       " 'between': 77,\n",
       " 'tissue': 78,\n",
       " 'ich': 79,\n",
       " 'cereus': 80,\n",
       " 'target': 81,\n",
       " 'assay': 82,\n",
       " 'database': 83,\n",
       " 'polymerase': 84,\n",
       " 'quantitative': 85,\n",
       " 'including': 86,\n",
       " 'factor': 87,\n",
       " 'abundance': 88,\n",
       " 'invasion': 89,\n",
       " 'striatal': 90,\n",
       " 'overexpression': 91,\n",
       " 'patients': 92,\n",
       " 'resistance': 93,\n",
       " 'tmao': 94,\n",
       " 'pathway': 95,\n",
       " 'crucial': 96,\n",
       " 'may': 97,\n",
       " 'btz': 98,\n",
       " 'mechanisms': 99,\n",
       " 'cancer': 100,\n",
       " 'observed': 101,\n",
       " 'markers': 102,\n",
       " 'concentrations': 103,\n",
       " 'be': 104,\n",
       " 'into': 105,\n",
       " 'trnaaspgtc': 106,\n",
       " 'stress': 107,\n",
       " 'studies': 108,\n",
       " 'positive': 109,\n",
       " 'glucosamine': 110,\n",
       " 'further': 111,\n",
       " 'its': 112,\n",
       " 'reference': 113,\n",
       " 'normal': 114,\n",
       " 'crc': 115,\n",
       " 'role': 116,\n",
       " 'under': 117,\n",
       " 'gbm': 118,\n",
       " 'culture': 119,\n",
       " 'lipid': 120,\n",
       " 'villosa': 121,\n",
       " 'diagnostic': 122,\n",
       " 'conclusion': 123,\n",
       " 'has': 124,\n",
       " 'species': 125,\n",
       " 'could': 126,\n",
       " 'data': 127,\n",
       " 'saltstressed': 128,\n",
       " 'stems': 129,\n",
       " 'coldstressed': 130,\n",
       " 'interaction': 131,\n",
       " 'mum': 132,\n",
       " 'model': 133,\n",
       " 'assessed': 134,\n",
       " 'western': 135,\n",
       " 'blotting': 136,\n",
       " 'determine': 137,\n",
       " 'both': 138,\n",
       " 'days': 139,\n",
       " 'level': 140,\n",
       " 'mouse': 141,\n",
       " 'controls': 142,\n",
       " 'dems': 143,\n",
       " 'regulation': 144,\n",
       " 'chain': 145,\n",
       " 'reaction': 146,\n",
       " 'rtqpcr': 147,\n",
       " 'measured': 148,\n",
       " 'mixedculturebased': 149,\n",
       " 'isolates': 150,\n",
       " 'important': 151,\n",
       " 'along': 152,\n",
       " 'changes': 153,\n",
       " 'detect': 154,\n",
       " 'however': 155,\n",
       " 'promoted': 156,\n",
       " 'have': 157,\n",
       " 'pmlkl': 158,\n",
       " 'pjnk': 159,\n",
       " 'zone': 160,\n",
       " 'increase': 161,\n",
       " 'activation': 162,\n",
       " 'single': 163,\n",
       " 'detected': 164,\n",
       " 'mexican': 165,\n",
       " 'vegf': 166,\n",
       " 'pten': 167,\n",
       " 'phylogenetically': 168,\n",
       " 'inflammation': 169,\n",
       " 'treatment': 170,\n",
       " 'differentiation': 171,\n",
       " 'not': 172,\n",
       " 'growth': 173,\n",
       " 'suppressed': 174,\n",
       " 'well': 175,\n",
       " 'tumor': 176,\n",
       " 'at': 177,\n",
       " 'targets': 178,\n",
       " 'vivo': 179,\n",
       " 'powder': 180,\n",
       " 'new': 181,\n",
       " 'downregulated': 182,\n",
       " 'damage': 183,\n",
       " 'trnavalaaccac': 184,\n",
       " 'chili': 185,\n",
       " 'but': 186,\n",
       " 'pregnancy': 187,\n",
       " 'materials': 188,\n",
       " 'used': 189,\n",
       " 'it': 190,\n",
       " 'samples': 191,\n",
       " 'mechanism': 192,\n",
       " 'dehydrogenase': 193,\n",
       " 'myeloma': 194,\n",
       " 'stage': 195,\n",
       " 'diet': 196,\n",
       " 'upregulated': 197,\n",
       " 'drought': 198,\n",
       " 'validated': 199,\n",
       " 'effects': 200,\n",
       " 'mice': 201,\n",
       " 'associated': 202,\n",
       " 'mirnas': 203,\n",
       " 'inhibiting': 204,\n",
       " 'group': 205,\n",
       " 'osmr': 206,\n",
       " 'cerebral': 207,\n",
       " 'kinase': 208,\n",
       " 'underlying': 209,\n",
       " 'drug': 210,\n",
       " 'models': 211,\n",
       " 'production': 212,\n",
       " 'clinical': 213,\n",
       " 'remain': 214,\n",
       " 'cancers': 215,\n",
       " 'exosomal': 216,\n",
       " 'human': 217,\n",
       " 'been': 218,\n",
       " 'additionally': 219,\n",
       " 'histone': 220,\n",
       " 'stability': 221,\n",
       " 'three': 222,\n",
       " 'leaves': 223,\n",
       " 'roots': 224,\n",
       " 'tested': 225,\n",
       " 'targeting': 226,\n",
       " 'assays': 227,\n",
       " 'besides': 228,\n",
       " 'iv': 229,\n",
       " 'then': 230,\n",
       " 'performed': 231,\n",
       " 'investigate': 232,\n",
       " 'sites': 233,\n",
       " 'transfected': 234,\n",
       " 'method': 235,\n",
       " 'roc': 236,\n",
       " 'curve': 237,\n",
       " 'sequencing': 238,\n",
       " 'respectively': 239,\n",
       " 'all': 240,\n",
       " 'pathways': 241,\n",
       " 'negative': 242,\n",
       " 'wb': 243,\n",
       " 'realtime': 244,\n",
       " 'proinflammatory': 245,\n",
       " 'evaluate': 246,\n",
       " 'experimental': 247,\n",
       " 'hours': 248,\n",
       " 'mtt': 249,\n",
       " 'apoptosis': 250,\n",
       " 'ischemiainduced': 251,\n",
       " 'pcr': 252,\n",
       " 'derived': 253,\n",
       " 'biochemical': 254,\n",
       " 'microgml': 255,\n",
       " 'hbl': 256,\n",
       " 'component': 257,\n",
       " 'range': 258,\n",
       " 'metastasis': 259,\n",
       " 'candidate': 260,\n",
       " 'enhancing': 261,\n",
       " 'metabolism': 262,\n",
       " 'values': 263,\n",
       " 'signaling': 264,\n",
       " 'macrophages': 265,\n",
       " 'factors': 266,\n",
       " 'indicate': 267,\n",
       " 'approach': 268,\n",
       " 'selected': 269,\n",
       " 'datasets': 270,\n",
       " 'composition': 271,\n",
       " 'viral': 272,\n",
       " 'mirna': 273,\n",
       " 'subventricular': 274,\n",
       " 'stromal': 275,\n",
       " 'differences': 276,\n",
       " 'cellular': 277,\n",
       " 'sl': 278,\n",
       " 'recovered': 279,\n",
       " 'future': 280,\n",
       " 'synthetase': 281,\n",
       " 'acid': 282,\n",
       " 'fragments': 283,\n",
       " 'accumulation': 284,\n",
       " 'regulate': 285,\n",
       " 'number': 286,\n",
       " 'trnaderived': 287,\n",
       " 'toxigenic': 288,\n",
       " 'noncoding': 289,\n",
       " 'possible': 290,\n",
       " 'necroptosis': 291,\n",
       " 'disease': 292,\n",
       " 'organism': 293,\n",
       " 'edema': 294,\n",
       " 'present': 295,\n",
       " 'condition': 296,\n",
       " 'characteristics': 297,\n",
       " 'neurobehavioral': 298,\n",
       " 'involved': 299,\n",
       " 'environment': 300,\n",
       " 'biomarker': 301,\n",
       " 'components': 302,\n",
       " 'sought': 303,\n",
       " 'while': 304,\n",
       " 'determined': 305,\n",
       " 'receptors': 306,\n",
       " 'during': 307,\n",
       " 'among': 308,\n",
       " 'thus': 309,\n",
       " 'virome': 310,\n",
       " 'macrophage': 311,\n",
       " 'high': 312,\n",
       " 'previously': 313,\n",
       " 'laryngocarcinoma': 314,\n",
       " 'prognostic': 315,\n",
       " 'structure': 316,\n",
       " 'elucidate': 317,\n",
       " 'response': 318,\n",
       " 'activity': 319,\n",
       " 'influence': 320,\n",
       " 'divergent': 321,\n",
       " 'silence': 322,\n",
       " 'various': 323,\n",
       " 'responsiveness': 324,\n",
       " 'regard': 325,\n",
       " 'pyruvate': 326,\n",
       " 'indicated': 327,\n",
       " 'region': 328,\n",
       " 'profiles': 329,\n",
       " 'metastatic': 330,\n",
       " 'some': 331,\n",
       " 'aim': 332,\n",
       " 'serum': 333,\n",
       " 'early': 334,\n",
       " 'recently': 335,\n",
       " 'unknown': 336,\n",
       " 'facilitate': 337,\n",
       " 'downregulating': 338,\n",
       " 'prognosis': 339,\n",
       " 'conditions': 340,\n",
       " 'largely': 341,\n",
       " 'knockdown': 342,\n",
       " 'process': 343,\n",
       " 'key': 344,\n",
       " 'stem': 345,\n",
       " 'abiotic': 346,\n",
       " 'neural': 347,\n",
       " 'four': 348,\n",
       " 'ischemic': 349,\n",
       " 'expressed': 350,\n",
       " 'transfer': 351,\n",
       " 'because': 352,\n",
       " 'cause': 353,\n",
       " 'metabolic': 354,\n",
       " 'bacillus': 355,\n",
       " 'association': 356,\n",
       " 'effect': 357,\n",
       " 'novel': 358,\n",
       " 'er': 359,\n",
       " 'plantarum': 360,\n",
       " 'cerevisiae': 361,\n",
       " 'animal': 362,\n",
       " 'bacteria': 363,\n",
       " 'inhibition': 364,\n",
       " 'induces': 365,\n",
       " 'food': 366,\n",
       " 'sporeforming': 367,\n",
       " 'inflammatory': 368,\n",
       " 'reactions': 369,\n",
       " 'research': 370,\n",
       " 'result': 371,\n",
       " 'taken': 372,\n",
       " 'bioactive': 373,\n",
       " 'are': 374,\n",
       " 'deacetylase': 375,\n",
       " 'via': 376,\n",
       " 'sponging': 377,\n",
       " 'mechanistically': 378,\n",
       " 'cancerinhibiting': 379,\n",
       " 'together': 380,\n",
       " 'nine': 381,\n",
       " 'based': 382,\n",
       " 'prior': 383,\n",
       " 'transcriptomic': 384,\n",
       " 'treatments': 385,\n",
       " 'simulating': 386,\n",
       " 'bioinformatics': 387,\n",
       " 'assessments': 388,\n",
       " 'family': 389,\n",
       " 'elongation': 390,\n",
       " 'act': 391,\n",
       " 'actin': 392,\n",
       " 'droughtstressed': 393,\n",
       " 'tua': 394,\n",
       " 'alphatubulin': 395,\n",
       " 'rrna': 396,\n",
       " 'utility': 397,\n",
       " 'mirnagene': 398,\n",
       " 'mode': 399,\n",
       " 'acylcoa': 400,\n",
       " 'oxidase': 401,\n",
       " 'lowest': 402,\n",
       " 'effective': 403,\n",
       " 'loecs': 404,\n",
       " 'silico': 405,\n",
       " 'molecule': 406,\n",
       " 'fluorescentbased': 407,\n",
       " 'electrophoretic': 408,\n",
       " 'mobility': 409,\n",
       " 'shift': 410,\n",
       " 'fremsa': 411,\n",
       " 'luciferase': 412,\n",
       " 'reporter': 413,\n",
       " 'directly': 414,\n",
       " 'bind': 415,\n",
       " 'inhibit': 416,\n",
       " 'clarify': 417,\n",
       " 'monocyte': 418,\n",
       " 'chemoattractant': 419,\n",
       " 'identify': 420,\n",
       " 'established': 421,\n",
       " 'injecting': 422,\n",
       " 'collagenase': 423,\n",
       " 'right': 424,\n",
       " 'basal': 425,\n",
       " 'ganglia': 426,\n",
       " 'inhibitor': 427,\n",
       " 'administered': 428,\n",
       " 'intraventricular': 429,\n",
       " 'injection': 430,\n",
       " 'employed': 431,\n",
       " 'phosphomixed': 432,\n",
       " 'lineage': 433,\n",
       " 'domainlike': 434,\n",
       " 'phosphocjun': 435,\n",
       " 'nterminal': 436,\n",
       " 'interleukin': 437,\n",
       " 'localization': 438,\n",
       " 'immunofluorescence': 439,\n",
       " 'staining': 440,\n",
       " 'coimmunoprecipitation': 441,\n",
       " 'sham': 442,\n",
       " 'found': 443,\n",
       " 'examination': 444,\n",
       " 'strongly': 445,\n",
       " 'implantation': 446,\n",
       " 'interimplantation': 447,\n",
       " 'location': 448,\n",
       " 'distributed': 449,\n",
       " 'decidual': 450,\n",
       " 'upon': 451,\n",
       " 'exploration': 452,\n",
       " 'artificially': 453,\n",
       " 'induced': 454,\n",
       " 'vitro': 455,\n",
       " 'interfering': 456,\n",
       " 'primary': 457,\n",
       " 'endometrial': 458,\n",
       " 'expressions': 459,\n",
       " 'uterus': 460,\n",
       " 'impaired': 461,\n",
       " 'led': 462,\n",
       " 'loss': 463,\n",
       " 'implanted': 464,\n",
       " 'embryos': 465,\n",
       " 'downloaded': 466,\n",
       " 'omnibus': 467,\n",
       " 'differentially': 468,\n",
       " 'limma': 469,\n",
       " 'receiver': 470,\n",
       " 'operating': 471,\n",
       " 'characteristic': 472,\n",
       " 'predicted': 473,\n",
       " 'mirwalk': 474,\n",
       " 'enrichment': 475,\n",
       " 'david': 476,\n",
       " 'tcga': 477,\n",
       " 'immunohistochemistry': 478,\n",
       " 'seven': 479,\n",
       " 'shared': 480,\n",
       " 'two': 481,\n",
       " 'constituted': 482,\n",
       " 'signature': 483,\n",
       " 'distinguish': 484,\n",
       " 'area': 485,\n",
       " 'nearly': 486,\n",
       " 'approximate': 487,\n",
       " 'ii': 488,\n",
       " 'promoter': 489,\n",
       " 'division': 490,\n",
       " 'dnatemplated': 491,\n",
       " 'cellcounting': 492,\n",
       " 'evaluating': 493,\n",
       " 'vascular': 494,\n",
       " 'endothelial': 495,\n",
       " 'quantified': 496,\n",
       " 'enzymelinked': 497,\n",
       " 'immunosorbent': 498,\n",
       " 'gl': 499,\n",
       " 'optimal': 500,\n",
       " 'concentration': 501,\n",
       " 'improve': 502,\n",
       " 'astrocytes': 503,\n",
       " 'oligodendrocytes': 504,\n",
       " 'meanwhile': 505,\n",
       " 'elevate': 506,\n",
       " 'transfection': 507,\n",
       " 'sirna': 508,\n",
       " 'stimulates': 509,\n",
       " 'cytokines': 510,\n",
       " 'designed': 511,\n",
       " 'tmaotreated': 512,\n",
       " 'exposed': 513,\n",
       " 'murine': 514,\n",
       " 'line': 515,\n",
       " 'butyric': 516,\n",
       " 'tunicamycin': 517,\n",
       " 'inducing': 518,\n",
       " 'erstress': 519,\n",
       " 'test': 520,\n",
       " 'oneway': 521,\n",
       " 'anova': 522,\n",
       " 'statistical': 523,\n",
       " 'no': 524,\n",
       " 'death': 525,\n",
       " 'doses': 526,\n",
       " 'significant': 527,\n",
       " 'elevation': 528,\n",
       " 'andor': 529,\n",
       " 'when': 530,\n",
       " 'completely': 531,\n",
       " 'aims': 532,\n",
       " 'antagomir': 533,\n",
       " 'if': 534,\n",
       " 'analysed': 535,\n",
       " 'colorimetric': 536,\n",
       " 'bromide': 537,\n",
       " 'annexin': 538,\n",
       " 'vpropidium': 539,\n",
       " 'iodide': 540,\n",
       " 'pi': 541,\n",
       " 'flow': 542,\n",
       " 'cytometry': 543,\n",
       " 'qrtpcr': 544,\n",
       " 'implemented': 545,\n",
       " 'bcell': 546,\n",
       " 'lymphoma': 547,\n",
       " 'antiapoptotic': 548,\n",
       " 'cycle': 549,\n",
       " 'exacerbated': 550,\n",
       " 'rate': 551,\n",
       " 'inadequacy': 552,\n",
       " 'existing': 553,\n",
       " 'developed': 554,\n",
       " 'mimicking': 555,\n",
       " 'internal': 556,\n",
       " 'artery': 557,\n",
       " 'occlusion': 558,\n",
       " 'mild': 559,\n",
       " 'moderate': 560,\n",
       " 'ameliorated': 561,\n",
       " 'behavioral': 562,\n",
       " 'dysfunction': 563,\n",
       " 'rescued': 564,\n",
       " 'upregulation': 565,\n",
       " 'ex': 566,\n",
       " 'neurosphere': 567,\n",
       " 'involving': 568,\n",
       " 'cellsneural': 569,\n",
       " 'progenitor': 570,\n",
       " 'large': 571,\n",
       " 'neurospheres': 572,\n",
       " 'nestin': 573,\n",
       " 'lastly': 574,\n",
       " 'coronal': 575,\n",
       " 'sixty': 576,\n",
       " 'postoperative': 577,\n",
       " 'consisting': 578,\n",
       " 'adjacent': 579,\n",
       " 'collected': 580,\n",
       " 'mintbase': 581,\n",
       " 'select': 582,\n",
       " 'taqman': 583,\n",
       " 'technology': 584,\n",
       " 'correlation': 585,\n",
       " 'clinicopathological': 586,\n",
       " 'lower': 587,\n",
       " 'comparison': 588,\n",
       " 'iii': 589,\n",
       " 'trnavalaac': 590,\n",
       " 'basic': 591,\n",
       " 'nutriments': 592,\n",
       " 'living': 593,\n",
       " 'development': 594,\n",
       " 'influencing': 595,\n",
       " 'healthdisease': 596,\n",
       " 'chromatin': 597,\n",
       " 'dna': 598,\n",
       " 'methylation': 599,\n",
       " 'modification': 600,\n",
       " 'fermented': 601,\n",
       " 'either': 602,\n",
       " 'mixed': 603,\n",
       " 'proteomics': 604,\n",
       " 'show': 605,\n",
       " 'carbohydrate': 606,\n",
       " 'greater': 607,\n",
       " 'llactate': 608,\n",
       " 'udpglucose': 609,\n",
       " 'contribute': 610,\n",
       " 'quality': 611,\n",
       " 'than': 612,\n",
       " 'those': 613,\n",
       " 'produced': 614,\n",
       " 'calreticulin': 615,\n",
       " 'enolase': 616,\n",
       " 'seryltrna': 617,\n",
       " 'needed': 618,\n",
       " 'proteins': 619,\n",
       " 'compounds': 620,\n",
       " 'play': 621,\n",
       " 'members': 622,\n",
       " 'colonies': 623,\n",
       " 'grew': 624,\n",
       " 'agar': 625,\n",
       " 'medium': 626,\n",
       " 'selective': 627,\n",
       " 'sensu': 628,\n",
       " 'lato': 629,\n",
       " 'supplemented': 630,\n",
       " 'polymyxin': 631,\n",
       " 'ampicillin': 632,\n",
       " 'presumptive': 633,\n",
       " 'trnacyspcrbased': 634,\n",
       " 'thuringiensis': 635,\n",
       " 'toyonensis': 636,\n",
       " 'serological': 637,\n",
       " 'tests': 638,\n",
       " 'enterotoxins': 639,\n",
       " 'nhe': 640,\n",
       " 'hemolysin': 641,\n",
       " 'bl': 642,\n",
       " 'c': 643,\n",
       " 'hblc': 644,\n",
       " 'nonhemolytic': 645,\n",
       " 'enterotoxin': 646,\n",
       " 'nhea': 647,\n",
       " 'antibiotic': 648,\n",
       " 'betalactams': 649,\n",
       " 'trimethoprimsulfamethoxazole': 650,\n",
       " 'tetracycline': 651,\n",
       " 'erythromycin': 652,\n",
       " 'clindamycin': 653,\n",
       " 'chloramphenicol': 654,\n",
       " 'korana': 655,\n",
       " 'river': 656,\n",
       " 'croatia': 657,\n",
       " 'highthroughput': 658,\n",
       " 'depleted': 659,\n",
       " 'total': 660,\n",
       " 'isolated': 661,\n",
       " 'hepatopancreas': 662,\n",
       " 'subsequent': 663,\n",
       " 'sequence': 664,\n",
       " 'crayfishassociated': 665,\n",
       " 'reolike': 666,\n",
       " 'hepelike': 667,\n",
       " 'totilike': 668,\n",
       " 'picornalike': 669,\n",
       " 'crustacean': 670,\n",
       " 'hosts': 671,\n",
       " 'patterns': 672,\n",
       " 'reads': 673,\n",
       " 'calculated': 674,\n",
       " 'nucleotide': 675,\n",
       " 'diversities': 676,\n",
       " 'sequences': 677,\n",
       " 'varied': 678,\n",
       " 'processes': 679,\n",
       " 'eg': 680,\n",
       " 'population': 681,\n",
       " 'density': 682,\n",
       " 'nonrandom': 683,\n",
       " 'dispersal': 684,\n",
       " 'host': 685,\n",
       " 'individuals': 686,\n",
       " 'core': 687,\n",
       " 'mimic': 688,\n",
       " 'specifically': 689,\n",
       " 'disorders': 690,\n",
       " 'pathogenesis': 691,\n",
       " 'due': 692,\n",
       " 'sections': 693,\n",
       " 'energy': 694,\n",
       " 'possibly': 695,\n",
       " 'containing': 696,\n",
       " 'presence': 697,\n",
       " 'isolation': 698,\n",
       " 'pparalphaindependent': 699,\n",
       " 'postich': 700,\n",
       " 'objective': 701,\n",
       " 'hypoxiainducible': 702,\n",
       " 'direct': 703,\n",
       " 'considerably': 704,\n",
       " 'general': 705,\n",
       " 'source': 706,\n",
       " 'nutritional': 707,\n",
       " 'fronts': 708,\n",
       " 'manner': 709,\n",
       " 'clpfesainduced': 710,\n",
       " 'pretreatment': 711,\n",
       " 'sirnamediated': 712,\n",
       " 'tlrs': 713,\n",
       " 'control': 714,\n",
       " 'detection': 715,\n",
       " 'deacetylation': 716,\n",
       " 'investigated': 717,\n",
       " 'outcomes': 718,\n",
       " 'unsatisfactory': 719,\n",
       " 'signalling': 720,\n",
       " 'within': 721,\n",
       " 'unfavorable': 722,\n",
       " 'reduced': 723,\n",
       " 'profile': 724,\n",
       " 'impact': 725,\n",
       " 'tolllike': 726,\n",
       " 'tunicamycintreated': 727,\n",
       " 'remained': 728,\n",
       " 'only': 729,\n",
       " 'cascades': 730,\n",
       " 'flavour': 731,\n",
       " 'invasive': 732,\n",
       " 'dramatically': 733,\n",
       " 'tensin': 734,\n",
       " 'therapeutic': 735,\n",
       " 'utilization': 736,\n",
       " 'globe': 737,\n",
       " 'homolog': 738,\n",
       " 'pattern': 739,\n",
       " 'will': 740,\n",
       " 'reported': 741,\n",
       " 'injuries': 742,\n",
       " 'cultured': 743,\n",
       " 'stroke': 744,\n",
       " 'saccharomyces': 745,\n",
       " 'nuclear': 746,\n",
       " 'downregulation': 747,\n",
       " 'deleted': 748,\n",
       " 'neurological': 749,\n",
       " 'betaoxidation': 750,\n",
       " 'atherosclerosis': 751,\n",
       " 'triggers': 752,\n",
       " 'ligand': 753,\n",
       " 'citrate': 754,\n",
       " 'around': 755,\n",
       " 'housekeeping': 756,\n",
       " 'essential': 757,\n",
       " 'binding': 758,\n",
       " 'synthase': 759,\n",
       " 'diets': 760,\n",
       " 'chromosome': 761,\n",
       " 'contributing': 762,\n",
       " 'accompanied': 763,\n",
       " 'fatty': 764,\n",
       " 'many': 765,\n",
       " 'distant': 766,\n",
       " 'fermentation': 767,\n",
       " 'lactiplantibacillus': 768,\n",
       " 'aldehyde': 769,\n",
       " 'l': 770,\n",
       " 'native': 771,\n",
       " 'ten': 772,\n",
       " 'better': 773,\n",
       " 'like': 774,\n",
       " 'cooccurring': 775,\n",
       " 'mitigated': 776,\n",
       " 'reticulum': 777,\n",
       " 'part': 778,\n",
       " 'brdu': 779,\n",
       " 'rnas': 780,\n",
       " 'suite': 781,\n",
       " 'droughtrelated': 782,\n",
       " 'revealed': 783,\n",
       " 'toxicity': 784,\n",
       " 'steps': 785,\n",
       " 'explored': 786,\n",
       " 'endoplasmic': 787,\n",
       " 'oncogenic': 788,\n",
       " 'more': 789,\n",
       " 'decarboxylase': 790,\n",
       " 'leniusculus': 791,\n",
       " 'combined': 792,\n",
       " 'candidates': 793,\n",
       " 'small': 794,\n",
       " 'whole': 795,\n",
       " 'pacifastacus': 796,\n",
       " 'markedly': 797,\n",
       " 'akt': 798,\n",
       " 'establish': 799,\n",
       " 'catalyzing': 800,\n",
       " 'progress': 801,\n",
       " 'plays': 802,\n",
       " 'decrease': 803,\n",
       " 'enzyme': 804,\n",
       " 'interest': 805,\n",
       " 'rats': 806,\n",
       " 'problem': 807,\n",
       " 'dcx': 808,\n",
       " 'demonstrated': 809,\n",
       " 'interactions': 810,\n",
       " 'spice': 811,\n",
       " 'animals': 812,\n",
       " 'preventing': 813,\n",
       " 'defects': 814,\n",
       " 'fetal': 815,\n",
       " 'major': 816,\n",
       " 'translation': 817,\n",
       " 'acetylcoa': 818,\n",
       " 'moreover': 819,\n",
       " 'provided': 820,\n",
       " 'worth': 821,\n",
       " 'low': 822,\n",
       " 'psammochloa': 823,\n",
       " 'first': 824,\n",
       " 'colocalized': 825,\n",
       " 'palmitoyltransferase': 826,\n",
       " 'noting': 827,\n",
       " 'obtained': 828,\n",
       " 'reveals': 829,\n",
       " 'report': 830,\n",
       " 'support': 831,\n",
       " 'carnitine': 832,\n",
       " 'given': 833,\n",
       " 'marker': 834,\n",
       " 'experiments': 835,\n",
       " 'hepatotoxicity': 836,\n",
       " 'exposure': 837,\n",
       " 'unclear': 838,\n",
       " 'remains': 839,\n",
       " 'summary': 840,\n",
       " 'selecting': 841,\n",
       " 'relevant': 842,\n",
       " 'perinuclear': 843,\n",
       " 'screen': 844,\n",
       " 'hippocampal': 845,\n",
       " 'newly': 846,\n",
       " 'genome': 847,\n",
       " 'organisms': 848,\n",
       " 'analyze': 849,\n",
       " 'tumors': 850,\n",
       " 'groups': 851,\n",
       " 'risk': 852,\n",
       " 'differentiating': 853,\n",
       " 'undiscovered': 854,\n",
       " 'validating': 855,\n",
       " 'aimed': 856,\n",
       " 'about': 857,\n",
       " 'diversity': 858,\n",
       " 'few': 859,\n",
       " 'might': 860,\n",
       " 'view': 861,\n",
       " 'induce': 862,\n",
       " 'ischemia': 863,\n",
       " 'immature': 864,\n",
       " 'known': 865,\n",
       " 'signally': 866,\n",
       " 'upregulate': 867,\n",
       " 'miscarriage': 868,\n",
       " 'neurons': 869,\n",
       " 'insights': 870,\n",
       " 'commercial': 871,\n",
       " 'decidua': 872,\n",
       " 'little': 873,\n",
       " 'infertility': 874,\n",
       " 'rarely': 875,\n",
       " 'weaker': 876,\n",
       " 'induction': 877,\n",
       " 'injury': 878,\n",
       " 'glucosaminetreated': 879,\n",
       " 'salt': 880,\n",
       " 'formation': 881,\n",
       " 'external': 882,\n",
       " 'modern': 883,\n",
       " 'dough': 884,\n",
       " 'wildlife': 885,\n",
       " 'linked': 886,\n",
       " 'hypoxia': 887,\n",
       " 'exact': 888,\n",
       " 'help': 889,\n",
       " 'raw': 890,\n",
       " 'annuum': 891,\n",
       " 'understudied': 892,\n",
       " 'oncogene': 893,\n",
       " 'cold': 894,\n",
       " 'environmental': 895,\n",
       " 'hemorrhage': 896,\n",
       " 'against': 897,\n",
       " 'cumulatively': 898,\n",
       " 'capsicum': 899,\n",
       " 'toxins': 900,\n",
       " 'intracerebral': 901,\n",
       " 'complex': 902,\n",
       " 'spontaneous': 903,\n",
       " 'biomarkers': 904,\n",
       " 'gbmrelated': 905,\n",
       " 'power': 906,\n",
       " 'similar': 907,\n",
       " 'bortezomib': 908,\n",
       " 'aggressive': 909,\n",
       " 'information': 910,\n",
       " 'aberrant': 911,\n",
       " 'abortion': 912,\n",
       " 'glioblastoma': 913,\n",
       " 'although': 914,\n",
       " 'lines': 915,\n",
       " 'diseases': 916,\n",
       " 'fungal': 917,\n",
       " 'elaborated': 918,\n",
       " 'functions': 919,\n",
       " 'disturbed': 920,\n",
       " 'what': 921,\n",
       " 'multiple': 922,\n",
       " 'marrow': 923,\n",
       " 'neuroprotective': 924,\n",
       " 'antineuroinflammatory': 925,\n",
       " 'proneurogenic': 926,\n",
       " 'confirming': 927,\n",
       " 'heterogeneous': 928,\n",
       " 'behavior': 929,\n",
       " 's': 930,\n",
       " 'pathogens': 931,\n",
       " 'sets': 932,\n",
       " 'sulfonate': 933,\n",
       " 'frequently': 934,\n",
       " 'pathological': 935,\n",
       " 'findings': 936,\n",
       " 'happens': 937,\n",
       " 'bone': 938,\n",
       " 'highly': 939,\n",
       " 'bacterial': 940,\n",
       " 'antibiotics': 941,\n",
       " 'foundations': 942,\n",
       " 'benefit': 943,\n",
       " 'homeostasis': 944,\n",
       " 'suggest': 945,\n",
       " 'leads': 946,\n",
       " 'constituents': 947,\n",
       " 'understand': 948,\n",
       " 'numerous': 949,\n",
       " 'conclusions': 950,\n",
       " 'step': 951,\n",
       " 'recommend': 952,\n",
       " 'understanding': 953,\n",
       " 'considerable': 954,\n",
       " 'investigations': 955,\n",
       " 'pfos': 956,\n",
       " 'there': 957,\n",
       " 'sugar': 958,\n",
       " 'malignant': 959,\n",
       " 'relations': 960,\n",
       " 'exerted': 961,\n",
       " 'plasma': 962,\n",
       " 'amino': 963,\n",
       " 'diagnosed': 964,\n",
       " 'finding': 965,\n",
       " 'consist': 966,\n",
       " 'activities': 967,\n",
       " 'successful': 968,\n",
       " 'abnormally': 969,\n",
       " 'china': 970,\n",
       " 'perfluorooctane': 971,\n",
       " 'following': 972,\n",
       " 'collectively': 973,\n",
       " 'considered': 974,\n",
       " 'provides': 975,\n",
       " 'neoplastic': 976,\n",
       " 'occurring': 977,\n",
       " 'commonly': 978,\n",
       " 'promising': 979,\n",
       " 'nutrigenomics': 980,\n",
       " 'serve': 981,\n",
       " 'microbiological': 982,\n",
       " 'northwest': 983,\n",
       " 'genomic': 984,\n",
       " 'documentation': 985,\n",
       " 'naturally': 986,\n",
       " 'allow': 987,\n",
       " 'lncrna': 988,\n",
       " 'product': 989,\n",
       " 'represent': 990,\n",
       " 'enhance': 991,\n",
       " 'nutrigenetics': 992,\n",
       " 'designing': 993,\n",
       " 'formed': 994,\n",
       " 'assessment': 995,\n",
       " 'ecosystems': 996,\n",
       " 'virus': 997,\n",
       " 'transcript': 998,\n",
       " 'growing': 999,\n",
       " 'alternative': 1000,\n",
       " ...}"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tokenizer.word_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "vocab_size = len(tokenizer.word_index) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1048"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vocab_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "y = to_categorical(y, num_classes=vocab_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_length = X.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "###LSTM Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = Sequential()\n",
    "model.add(Embedding(vocab_size, 50, input_length=seq_length))\n",
    "model.add(LSTM(100, return_sequences=True))\n",
    "model.add(LSTM(100))\n",
    "model.add(Dense(100, activation='relu'))\n",
    "model.add(Dense(vocab_size, activation='softmax'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"sequential\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " embedding (Embedding)       (None, 50, 50)            52400     \n",
      "                                                                 \n",
      " lstm (LSTM)                 (None, 50, 100)           60400     \n",
      "                                                                 \n",
      " lstm_1 (LSTM)               (None, 100)               80400     \n",
      "                                                                 \n",
      " dense (Dense)               (None, 100)               10100     \n",
      "                                                                 \n",
      " dense_1 (Dense)             (None, 1048)              105848    \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 309,148\n",
      "Trainable params: 309,148\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.compile(loss = 'categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/30\n",
      "9/9 [==============================] - 6s 208ms/step - loss: 6.9436 - accuracy: 0.0429\n",
      "Epoch 2/30\n",
      "9/9 [==============================] - 2s 193ms/step - loss: 6.5846 - accuracy: 0.0318\n",
      "Epoch 3/30\n",
      "9/9 [==============================] - 2s 198ms/step - loss: 6.0632 - accuracy: 0.0407\n",
      "Epoch 4/30\n",
      "9/9 [==============================] - 2s 207ms/step - loss: 5.9265 - accuracy: 0.0486\n",
      "Epoch 5/30\n",
      "9/9 [==============================] - 2s 212ms/step - loss: 5.8819 - accuracy: 0.0535\n",
      "Epoch 6/30\n",
      "9/9 [==============================] - 3s 286ms/step - loss: 5.8666 - accuracy: 0.0531\n",
      "Epoch 7/30\n",
      "9/9 [==============================] - 2s 251ms/step - loss: 5.8569 - accuracy: 0.0553\n",
      "Epoch 8/30\n",
      "9/9 [==============================] - 2s 259ms/step - loss: 5.8539 - accuracy: 0.0579\n",
      "Epoch 9/30\n",
      "9/9 [==============================] - 2s 219ms/step - loss: 5.8531 - accuracy: 0.0579\n",
      "Epoch 10/30\n",
      "9/9 [==============================] - 2s 237ms/step - loss: 5.8516 - accuracy: 0.0579\n",
      "Epoch 11/30\n",
      "9/9 [==============================] - 2s 264ms/step - loss: 5.8501 - accuracy: 0.0579\n",
      "Epoch 12/30\n",
      "9/9 [==============================] - 2s 237ms/step - loss: 5.8498 - accuracy: 0.0579\n",
      "Epoch 13/30\n",
      "9/9 [==============================] - 2s 232ms/step - loss: 5.8532 - accuracy: 0.0579\n",
      "Epoch 14/30\n",
      "9/9 [==============================] - 3s 281ms/step - loss: 5.8502 - accuracy: 0.0544\n",
      "Epoch 15/30\n",
      "9/9 [==============================] - 2s 259ms/step - loss: 5.8473 - accuracy: 0.0579\n",
      "Epoch 16/30\n",
      "9/9 [==============================] - 3s 277ms/step - loss: 5.8458 - accuracy: 0.0579\n",
      "Epoch 17/30\n",
      "9/9 [==============================] - 3s 288ms/step - loss: 5.8464 - accuracy: 0.0495\n",
      "Epoch 18/30\n",
      "9/9 [==============================] - 3s 280ms/step - loss: 5.8440 - accuracy: 0.0579\n",
      "Epoch 19/30\n",
      "9/9 [==============================] - 2s 270ms/step - loss: 5.8404 - accuracy: 0.0579\n",
      "Epoch 20/30\n",
      "9/9 [==============================] - 2s 267ms/step - loss: 5.8376 - accuracy: 0.0513\n",
      "Epoch 21/30\n",
      "9/9 [==============================] - 2s 253ms/step - loss: 5.8316 - accuracy: 0.0579\n",
      "Epoch 22/30\n",
      "9/9 [==============================] - 3s 290ms/step - loss: 5.8169 - accuracy: 0.0597\n",
      "Epoch 23/30\n",
      "9/9 [==============================] - 3s 290ms/step - loss: 5.7936 - accuracy: 0.0579\n",
      "Epoch 24/30\n",
      "9/9 [==============================] - 3s 306ms/step - loss: 5.7529 - accuracy: 0.0672\n",
      "Epoch 25/30\n",
      "9/9 [==============================] - 3s 303ms/step - loss: 5.6957 - accuracy: 0.0721\n",
      "Epoch 26/30\n",
      "9/9 [==============================] - 2s 263ms/step - loss: 5.6406 - accuracy: 0.0729\n",
      "Epoch 27/30\n",
      "9/9 [==============================] - 2s 249ms/step - loss: 5.5901 - accuracy: 0.0729\n",
      "Epoch 28/30\n",
      "9/9 [==============================] - 2s 279ms/step - loss: 5.5523 - accuracy: 0.0787\n",
      "Epoch 29/30\n",
      "9/9 [==============================] - 2s 267ms/step - loss: 5.5243 - accuracy: 0.0747\n",
      "Epoch 30/30\n",
      "9/9 [==============================] - 2s 232ms/step - loss: 5.4954 - accuracy: 0.0796\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.callbacks.History at 0x18b5f301e80>"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.fit(X, y, batch_size =256, epochs =30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
